### PR TITLE
connectrpc: generate code with buf

### PR DESCRIPTION
**Pull request by oncilla**
Generate the additional code required for connectrpc with the buf tool. Currently, there is no bazel ruleset available for buf. We work around this by invoking the buf binary directly and commiting the generated code in git. To deal with generation, we pin the `protoc-gen-connect-go` version in the go.mod file. A small scaffolding go binary is added to deal with setting up the environment correctly for the buf tool. While this could be a shell script, this is more readable and maintainable.

This PR is broken out of the
https://github.com/scionproto/scion/tree/connectrpc branch to make incremental review easier.

TL;DR: Run `bazel run //tools/buf` to generate the connectrpc code.

Contributes to #4434

### CODE REVIEW

**Comment by matzf** [comment]
This change is [Reviewable](https://reviewable.io/reviews/scionproto/scion/4540)

**Review by matzf — COMMENTED** [review]
> Currently, there is no bazel ruleset available for buf.

Couldn't we use the connect plugin for protoc directly of using buf? I saw some mentions of plugins in the documentation of the bazel protobuf rules that we use (rules_proto_grpc). Does this not work, too complicated, or other reasons to prefer buf?


> While this could be a shell script, this is more readable and maintainable.

I know this is very much opinion based, but I personally do not find this go program is more readable than an equivalent shell script. It only does things that shell excells at (massaging paths, invoking a binary).

> Reviewed 27 of 28 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4540#-:-O-GI9XD9kD1-b9jlryF:b-t0cix5)* status: 27 of 28 files reviewed, 7 unresolved discussions (waiting on @jiceatscion and @oncilla)

___
*[`Makefile` line 84 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-7axX0Dv42xoQiVDr7:-O-7axX0Dv42xoQiVDr8:b5e1e4o) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/Makefile#L84)):*
> ```Makefile
> 	chmod 0644 pkg/proto/*/*.pb.go
> 
> 	bazel run //tools/buf
> ```

Clean files first to make sure no stale files are left?
`rm -f pkg/proto/*/v1/*connect/*.connect.go`
___
*[`pkg/proto/control_plane/v1/control_planeconnect/cppki.connect.go` line 39 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-GQnl74vQFt_W4NEPU:-O-GQnl74vQFt_W4NEPV:biiavw3) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/pkg/proto/control_plane/v1/control_planeconnect/cppki.connect.go#L39)):*
> ```Go
> const (
> 	// TrustMaterialServiceName is the fully-qualified name of the TrustMaterialService service.
> 	TrustMaterialServiceName = "proto.control_plane.v1.TrustMaterialService"
> ```

We talked about "fixing" this namespace also in the same go, as we can use a different "package" in connect than in grpc.
Do you think this could be addressed already in this code generation step, or should we do this later?
___
*[`private/app/xbazel/xbazel.go` line 27 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-6LEvT-1UNHZt-Fima:-O-6LEvT-1UNHZt-Fimb:b-kkke5) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/private/app/xbazel/xbazel.go#L27)):*
> ```Go
> )
> 
> const binDir = "bazel-out/k8-fastbuild/bin/"
> ```

Nit: Possibly unreliable. The `k8` is not fixed, it's the value of the `--cpu` flag; `fastbuild` is the `--compilation_mode`. 
But see comment below in `resolveFile`, I believe `binDir` is not necessary.
___
*[`private/app/xbazel/xbazel.go` line 65 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-GOFfB1DKW_1cw7XyM:-O-GOFfB1DKW_1cw7XyN:b-2pheim) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/private/app/xbazel/xbazel.go#L65)):*
> ```Go
> 	// XXX: Strip the bin directory from the file path. This is required for
> 	// some go binaries.
> 	if strings.HasPrefix(filepath.ToSlash(file), binDir) {
> ```

Not entirely sure, but this workaround can be removed if we use `$(rootpath ...)` instead of `(location @com_connectrpc_connect//cmd/protoc-gen-connect-go)`. `rootpath` should result in a path relative to runfiles.  
___
*[`private/app/xbazel/xbazel.go` line 76 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-GPVDbDY_PLjxHuz8_:-O-GPVDbDY_PLjxHuz8a:brric8w) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/private/app/xbazel/xbazel.go#L76)):*
> ```Go
> 	}
> 
> 	abs, err := filepath.Abs(file)
> ```

Isn't this a nop after `filepath.Join(runfilesDir, file)`? Or maybe `filepath.Clean`?
___
*[`tools/buf/BUILD.bazel` line 16 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-6WC2T0N2_8RtVyZjL:-O-6WC2T0N2_8RtVyZjM:bf6dlof) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/tools/buf/BUILD.bazel#L16)):*
> ```Starlark
> 
> go_binary(
>     name = "buf",
> ```

Somewhat of a nitpick; this script here always invokes `buf generate` with specific flags; it does not allow passing any flags or other parameters to `buf`. I think this should be reflected in the name of the binary and the bazel target, e.g. call it `buf-generate-connect-go` or so. 
___
*[`tools/buf/main.go` line 50 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-6WJNj7nqdSyS_BhC_:-O-6WJNj7nqdSyS_BhCa:bfq75z4) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/tools/buf/main.go#L50)):*
> ```Go
> 	for _, file := range must.Get(filepath.Glob("bazel-*")) {
> 		args = append(args, "--exclude-path", file)
> 	}
> ```

On my machine, this still seems to get hung up in some symlink trap. It works fine when I change this to pass `--path`, `proto/` instead of the `--exclude-path`s.

**Review by jiceatscion — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4540)* status: 27 of 28 files reviewed, 8 unresolved discussions (waiting on @matzf and @oncilla)

___
*[`tools/buf/main.go` line 14 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-Kuk7kEkRW7SSUvgrL:-O-Kuk7kEkRW7SSUvgrM:boyebi1) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/tools/buf/main.go#L14)):*
> ```Go
> // See the License for the specific language governing permissions and
> // limitations under the License.
> ```

I really don't see the benefit of having this go program just to invoke one executable with a trivial template. As it is, that's 4 extra moving parts just to run a simple command.

**Review by oncilla — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4540)* status: 27 of 28 files reviewed, 8 unresolved discussions (waiting on @jiceatscion and @matzf)

___
*[`tools/buf/main.go` line 14 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-Kuk7kEkRW7SSUvgrL:-O-M6uNzBRa4Gq-7fNFA:b-o8g1nq) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/tools/buf/main.go#L14)):*
<details><summary><i>Previously, jiceatscion wrote…</i></summary><blockquote>

I really don't see the benefit of having this go program just to invoke one executable with a trivial template. As it is, that's 4 extra moving parts just to run a simple command.
</blockquote></details>

Every single bash script I have written for bazel rules/macros I have regretted after some time.
This is my personal bias of course.

If you feel really strongly about it, we can drop it.
If you do not feel strongly, I would really prefer to keep it anyhing other than bash.

**Review by oncilla — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4540)* status: 27 of 28 files reviewed, 9 unresolved discussions (waiting on @jiceatscion and @matzf)

*[a discussion](https://reviewable.io/reviews/scionproto/scion/4540#-O0QR2CpD_mZVpJ5j3yw:-O0QR2CpD_mZVpJ5j3yx:b8jjeyf) (no related file):*
Reminder to self: https://github.com/abitofhelp/connect-go-example

___

**Review by jiceatscion — COMMENTED** [review]
> Reviewed 1 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4540)* status: all files reviewed, 8 unresolved discussions (waiting on @matzf and @oncilla)

___
*[`tools/buf/main.go` line 14 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-Kuk7kEkRW7SSUvgrL:-O5IuKY86kou1f1mZ6vF:bdy5h2w) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/tools/buf/main.go#L14)):*
<details><summary><i>Previously, oncilla (Dominik Roos) wrote…</i></summary><blockquote>

Every single bash script I have written for bazel rules/macros I have regretted after some time.
This is my personal bias of course.

If you feel really strongly about it, we can drop it.
If you do not feel strongly, I would really prefer to keep it anyhing other than bash.
</blockquote></details>

I do not feel strongly enough about it to block this PR.

**Review by oncilla — COMMENTED** [review]
> Dismissed @matzf from 7 discussions.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4540)* status: 3 of 33 files reviewed, all discussions resolved (waiting on @jiceatscion and @matzf)

___
*[`tools/buf/main.go` line 14 at r1](https://reviewable.io/reviews/scionproto/scion/4540#-O-Kuk7kEkRW7SSUvgrL:-O6mMNeG9ig7ZnryeEFK:bd6oyuh) ([raw file](https://github.com/scionproto/scion/blob/1aa57f9550d60b988ea6a47b2bd447a975c60de0/tools/buf/main.go#L14)):*
<details><summary><i>Previously, jiceatscion wrote…</i></summary><blockquote>

I do not feel strongly enough about it to block this PR.
</blockquote></details>

Gone now anyway

**Review by oncilla — COMMENTED** [review]
> Reviewed 3 of 28 files at r1, 30 of 30 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4540)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @oncilla)

**Review by jiceatscion — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4540)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @oncilla)
